File: src/java/src/main/java/org/elastos/hive/ServiceEndpoint.java
package org.elastos.hive;

import java.io.File;
import java.lang.ref.WeakReference;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;

import org.elastos.did.jwt.Claims;
import org.elastos.did.jwt.JwtParserBuilder;
import org.elastos.hive.connection.NodeRPCConnection;
import org.elastos.hive.connection.NodeRPCException;
import org.elastos.hive.connection.auth.AccessToken;
import org.elastos.hive.connection.auth.BridgeHandler;
import org.elastos.hive.endpoint.AboutController;
import org.elastos.hive.endpoint.NodeVersion;
import org.elastos.hive.exception.HiveException;
import org.elastos.hive.exception.NotImplementedException;

public class ServiceEndpoint extends NodeRPCConnection {
	private AppContext context;
	private String providerAddress;

	private String appDid;
	private String appInstanceDid;
	private String serviceInstanceDid;

	private AccessToken accessToken;
	private DataStorage dataStorage;

	protected ServiceEndpoint(AppContext context, String providerAddress) {
		if (context == null || providerAddress == null)
			throw new IllegalArgumentException("Empty context or provider address parameter");

		this.context = context;
		this.providerAddress = providerAddress;

		String dataDir = context.getAppContextProvider().getLocalDataDir();
		if (!dataDir.endsWith(File.separator))
			dataDir += File.separator;

		this.dataStorage = new FileStorage(dataDir, context.getUserDid());
		this.accessToken = new AccessToken(this, dataStorage, new BridgeHandler() {
			private WeakReference<ServiceEndpoint> weakref;

			@Override
			public void flush(String value) {
				try {
					ServiceEndpoint endpoint = weakref.get();
					Claims claims;

					claims = new JwtParserBuilder().build().parseClaimsJws(value).getBody();
					endpoint.flushDids(claims.getAudience(), claims.getIssuer());

				} catch (Exception e) {
					e.printStackTrace();
					return;
				}
			}

			BridgeHandler setTarget(ServiceEndpoint endpoint) {
				this.weakref = new WeakReference<>(endpoint);
				return this;
			}

			@Override
			public Object target() {
				return weakref.get();
			}

		}.setTarget(this));
	}

	public AppContext getAppContext() {
		return context;
	}

	/**
	 * Get the end-point address of this service End-point.
	 *
	 * @return provider address
	 */
	@Override
	public String getProviderAddress() {
		return providerAddress;
	}

	/**
	 * Get the user DID string of this serviceEndpoint.
	 *
	 * @return user did
	 */
	public String getUserDid() {
		return context.getUserDid();
	}

	/**
	 * Get the application DID in the current calling context.
	 *
	 * @return application did
	 */
	public String getAppDid() {
		return appDid;
	}

	/**
	 * Get the application instance DID in the current calling context;
	 *
	 * @return application instance did
	 */
	public String getAppInstanceDid() {
		return appInstanceDid;
	}

	/**
	 * Get the remote node service application DID.
	 *
	 * @return node service did
	 */
	public String getServiceDid() {
		throw new NotImplementedException();
	}

	/**
	 * Get the remote node service instance DID where is serving the storage service.
	 *
	 * @return node service instance did
	 */
	public String getServiceInstanceDid() {
		return serviceInstanceDid;
	}

	private void flushDids(String appInstanceDId, String serviceInstanceDid) {
		this.appInstanceDid = appInstanceDId;
		this.serviceInstanceDid = serviceInstanceDid;
	}

	public DataStorage getStorage() {
		return dataStorage;
	}

	public void refreshAccessToken() throws NodeRPCException {
		accessToken.fetch();
	}

	@Override
	protected AccessToken getAccessToken() {
		return accessToken;
	}

	public CompletableFuture<NodeVersion> getNodeVersion() {
		return CompletableFuture.supplyAsync(() -> {
			try {
				return new AboutController(this).getNodeVersion();
			} catch (HiveException | RuntimeException e) {
				throw new CompletionException(e);
			}
		});
	}

	public CompletableFuture<String> getLatestCommitId() {
		return CompletableFuture.supplyAsync(() -> {
			try {
				return new AboutController(this).getCommitId();
			} catch (HiveException | RuntimeException e) {
				throw new CompletionException(e);
			}
		});
	}
}

File: src/connection/serviceendpoint.ts
import {File} from '../utils/storage/file'
import {AppContext} from './auth/appcontext'
import {AccessToken} from './auth/accesstoken'
import {DataStorage} from '../utils/storage/datastorage'
import {FileStorage} from '../utils/storage/filestorage'
import {NodeVersion} from '../service/about/nodeversion'
import {HttpClient} from './httpclient'
import {AboutService} from '../service/about/aboutservice'
import {Logger} from '../utils/logger'
import {NodeInfo} from "../service/about/nodeinfo";

export class ServiceEndpoint {
    private readonly context: AppContext;
    private providerAddress: string;
    private aboutService: AboutService;
    private aboutServiceAuth: AboutService;
    private appInstanceDid: string;
    private serviceInstanceDid: string;

    private accessToken: AccessToken;
    private dataStorage: DataStorage;

    private static LOG_SERVICE_CONTEXT = new Logger("ServiceEndpoint");

    constructor(context: AppContext, providerAddress?: string) {
        if (!context && !providerAddress)
            throw new Error('Invalid parameter: context and providerAddress can not all empty');

        this.context = context;  // nullable
        this.providerAddress = providerAddress;

        this.init();
    }

    private init(): void {
        this.aboutService = new AboutService(this, new HttpClient(this, HttpClient.NO_AUTHORIZATION, HttpClient.DEFAULT_OPTIONS));

        if (this.context) {
            let dataDir = this.context.getAppContextProvider().getLocalDataDir();
            if (!dataDir.endsWith(File.SEPARATOR))
                dataDir += File.SEPARATOR;

            this.dataStorage = new FileStorage(dataDir, this.context.getUserDid());
            this.accessToken = new AccessToken(this, this.dataStorage);
            this.aboutServiceAuth = new AboutService(this, new HttpClient(this, HttpClient.WITH_AUTHORIZATION, HttpClient.DEFAULT_OPTIONS));
        }
    }

    hasAppContext(): boolean {
        return this.context != null;
    }

    getAccessToken(): AccessToken {
        if (!this.context)
            throw new Error('AppContext not setup');

        return this.accessToken;
    }

    async getProviderAddress(): Promise<string> {
        if (!this.providerAddress) {
            this.providerAddress = await this.context.getProviderAddress();
        }
        return this.providerAddress;
    }

    getAppContext(): AppContext {
        if (!this.context)
            throw new Error('AppContext not setup');

        return this.context;
    }

    /**
     * Get the user DID string of this ServiceEndpoint.
     *
     * @return user did
     */
    getUserDid(): string {
        if (!this.context)
            throw new Error('AppContext not setup');

        return this.context.getUserDid();
    }

    /**
     * Get the application DID in the current calling context.
     *
     * @return application did
     */
    getAppDid(): string {
        if (!this.context)
            throw new Error('AppContext not setup');

        return this.context.getAppDid();
    }

    /**
     * Get the application instance DID in the current calling context;
     *
     * @return application instance did
     */
    getAppInstanceDid(): string {
        return this.appInstanceDid;
    }

    /**
     * Get the remote node service application DID.
     *
     * @return node service did
     */
    getServiceDid(): string {
        return this.serviceInstanceDid;
    }

    /**
     * Get the remote node service instance DID where is serving the storage service.
     *
     * @return node service instance did
     */
    async getServiceInstanceDid(): Promise<string> {
        if (!this.serviceInstanceDid) {
            if (!this.context)
                throw new Error('AppContext not setup');

            await this.accessToken.fetch();
        }
        return this.serviceInstanceDid;
    }

    flushDids(appInstanceDId: string, serviceInstanceDid: string): void {
        this.appInstanceDid = appInstanceDId;
        this.serviceInstanceDid = serviceInstanceDid;
    }

    getStorage(): DataStorage {
        if (!this.context)
            throw new Error('AppContext not setup');

        return this.dataStorage;
    }

    async getNodeVersion(): Promise<NodeVersion> {
        return await this.aboutService.getNodeVersion();
    }

    async getLatestCommitId(): Promise<string> {
        return await this.aboutService.getCommitId();
    }

    async getNodeInfo(): Promise<NodeInfo> {
        if (!this.context)
            throw new Error('AppContext not setup');

        return await this.aboutServiceAuth.getInfo();
    }
}
